File: src/main/java/upbrella/be/umbrella/repository/UmbrellaRepositoryCustom.java
package upbrella.be.umbrella.repository;

import org.springframework.data.domain.Pageable;
import upbrella.be.umbrella.dto.response.UmbrellaWithHistory;

import java.util.List;

public interface UmbrellaRepositoryCustom {

    long countAllUmbrellas();

    long countRentableUmbrellas();

    long countRentedUmbrellas();

    long countMissingUmbrellas();

    long countRentableUmbrellasByStore(long storeMetaId);

    long countRentedUmbrellasByStore(long storeMetaId);

    long countAllUmbrellasByStore(long storeId);

    long countMissingUmbrellasByStore(long storeId);

    List<UmbrellaWithHistory> findUmbrellaAndHistoryOrderedByUmbrellaId(Pageable pageable);

    List<UmbrellaWithHistory> findUmbrellaAndHistoryOrderedByUmbrellaIdByStoreId(long storeId, Pageable pageable);
}

File: src/main/kotlin/upbrella/be/umbrella/repository/UmbrellaRepositoryCustom.kt
package upbrella.be.umbrella.repository

import org.springframework.data.domain.Pageable
import upbrella.be.umbrella.dto.response.UmbrellaWithHistory

interface UmbrellaRepositoryCustom {

    fun countAllUmbrellas(): Long

    fun countRentableUmbrellas(): Long

    fun countRentedUmbrellas(): Long

    fun countMissingUmbrellas(): Long

    fun countRentableUmbrellasByStore(storeMetaId: Long): Long

    fun countRentedUmbrellasByStore(storeMetaId: Long): Long

    fun countAllUmbrellasByStore(storeId: Long): Long

    fun countMissingUmbrellasByStore(storeId: Long): Long

    fun findUmbrellaAndHistoryOrderedByUmbrellaId(pageable: Pageable): List<UmbrellaWithHistory>

    fun findUmbrellaAndHistoryOrderedByUmbrellaIdByStoreId(storeId: Long, pageable: Pageable): List<UmbrellaWithHistory>
}